File: safmq.net/safmq.net.test/SafmqSSLTest.cs
using safmq.net;
using MbUnit.Framework;
using System;
using System.Collections.Generic;
using System.Security.Cryptography.X509Certificates;
using System.Net.Security;
using System.Net;
using System.Security.Authentication;

namespace safmq.net.test
{
    [TestFixture]
    public class SafmqSSLTest : SafmqTest
    {
        public static bool ServerValidator(object sender,
                                     X509Certificate certificate,
                                     X509Chain chain,
                                     SslPolicyErrors sslPolicyErrors) {
            Console.WriteLine("Server Certificate: " + certificate.Issuer);
            Console.WriteLine("SslPolicyErrors: " + sslPolicyErrors);
            return true;
        }

        public static X509Certificate CertificateProvider(object sender,
                                    string targetHost,
                                    X509CertificateCollection localCertificates,
                                    X509Certificate remoteCertificate,
                                    string[] acceptableIssuers) {
            // TODO: Build the CertificateProvider so the tests will not fail            
            try {
                //X509Certificate c = new X509Certificate(@"e:\blacksheep\safmq.0.8\safmq.net\safmq.net.test\sample.p12", "test");
                //Console.WriteLine("Issuer: " + c.Issuer);
                //Console.WriteLine("Subject: " + c.Subject);
                //foreach (string issuer in acceptableIssuers) {
                //    Console.WriteLine("Acceptable Issuer: " + issuer);
                //}
                if (localCertificates.Count > 0) {
                    Console.WriteLine("Offering client certificate: " + localCertificates[0].Subject);
                    return localCertificates[0];
                }
            } catch (Exception e) {
                Console.WriteLine(e.ToString());
            }
            return null;
        }

        string user = "npuser";
        string password = "54321";
        string description = "passwordless login test user";
        string subject = "OU=Testing, O=Safmq Java Test Client";
        string issuer = "C=US, ST=Nebraska, L=Omaha, O=SAFMQ, OU=Sample X509 Certificate, CN=Sample SAFMQ X509 Certificate";

        public SafmqSSLTest() {
            address = new Uri("safmqs://admin:@localhost:40001");
            simpleAddress = new Uri("safmqs://localhost:40001");
        }

        MQConnection connectToServer(Uri uri, string user, string password) {
            try {
                SSLContext ctx = new SSLContext();
                ctx.ServerValidator = SafmqSSLTest.ServerValidator;

                MQBuilder.ctx = ctx;

                MQConnection con = MQBuilder.BuildConnection(uri, user, password);
                Assert.IsTrue(con != null, "Successfully Connected");
                return con;
            } catch (Exception e) {
                Assert.IsTrue(e == null, e.ToString());
            }
            return null;
        }

        SSLContext setupSSL() {
            // Setup the SSL Context using the JKS file created via gen_test_cert in the java/cert directory
            MQConnection con = connectToServer(address, "", "");

            ErrorCode ec = con.DeleteUser(user);
            Assert.IsTrue(ec == ErrorCode.EC_NOERROR || ec == ErrorCode.EC_DOESNOTEXIST, "Delete User");
            ec = con.CreateUser(user, password, description);
            Assert.IsTrue(ec == ErrorCode.EC_NOERROR, "Create User");
            Console.WriteLine("Adding Identity, user: " + user + " subject: " + subject);
            ec = con.AddUserIdentity(subject, issuer, user);
            Assert.IsTrue(ec == ErrorCode.EC_NOERROR, "AddUserIdentity");

            con.Close();

            SSLContext ctx = new SSLContext();
            ctx.ServerValidator = new RemoteCertificateValidationCallback(SafmqSSLTest.ServerValidator);
            ctx.CertificateProvider = new LocalCertificateSelectionCallback(SafmqSSLTest.CertificateProvider);

            X509Certificate c = new X509Certificate(@"e:\blacksheep\safmq.trunk\safmq.net\safmq.net.test\sample.p12", "test");
            ctx.ClientCertificateList = new X509CertificateCollection();
            ctx.ClientCertificateList.Add(c);

            X509Certificate2 c2 = new X509Certificate2(@"e:\blacksheep\safmq.trunk\safmq.net\safmq.net.test\sample.p12", "test");
            X509Store store = new X509Store(StoreName.My);
            store.Open(OpenFlags.ReadWrite);
            store.Add(c2);
            store.Close();

            return ctx;
        }


        /**
         * Creates a user and associates an x.509 certificate with that user. Then
         * attempts to login without a password for that user.
         */
        [Test]
        [TestSequence(20)]
        public void test_PasswordlessSignon() {
            ErrorCode ec;

            MQConnection con = connectToServer(address, "", "");

            ec = con.DeleteUser(user);
            Assert.IsTrue(ec == ErrorCode.EC_NOERROR || ec == ErrorCode.EC_DOESNOTEXIST, "Delete User");

            ec = con.CreateUser(user, password, description);
            Assert.IsTrue(ec == ErrorCode.EC_NOERROR, "Create User");

            ec = con.AddUserIdentity(subject, issuer, user);
            Assert.IsTrue(ec == ErrorCode.EC_NOERROR, "AddUserIdentity");

            List<X509Identity> ids = new List<X509Identity>();
            ec = con.EnumerateUserIdentities(user, ids);
            Assert.IsTrue(ec == ErrorCode.EC_NOERROR, "EnumerateUserIdentities");
            Assert.IsTrue(ids.Count > 0, "X.509 ID Count");
            // TODO: Check the results in the ids
            bool idFound = false;
            for (int x = 0; x < ids.Count; x++) {
                X509Identity id = ids[0];
                if (id.IssuerDN.Equals(issuer) && id.SubjectDN.Equals(subject))
                    idFound = true;
            }
            Assert.IsTrue(idFound, "Identity found in returned X.509 identities");
            con.Close();
        }


        [Test]
        [TestSequence(21)]
        public void test_NullLogin() {
            SSLContext ctx = setupSSL();
            MQConnection con;

            try {
                con = MQBuilder.BuildConnection(simpleAddress, null, null, ctx);
                con.Close();
            } catch (Exception e) {
                Assert.IsTrue(e == null, "Reconnect with x.509 ident null ident " + e.ToString());
            }
        }

        [Test]
        [TestSequence(22)]
        public void test_EmptyLogin() {
            SSLContext ctx = setupSSL();
            MQConnection con;

            try {
                con = MQBuilder.BuildConnection(simpleAddress, "", "", ctx);
                con.Close();
            } catch (Exception e) {
                Assert.IsTrue(e == null, "Connect attempt with x.509 ident empty string " + e.ToString());
            }
        }

        [Test]
        [TestSequence(23)]
        public void test_NullPassword() {
            SSLContext ctx = setupSSL();
            MQConnection con;

            try {
                con = MQBuilder.BuildConnection(simpleAddress, user, null, ctx);
                con.Close();
            } catch (Exception e) {
                Assert.IsTrue(e == null, "Connect attempt with x.509 ident name, null password " + e.ToString());
            }

        }

        [Test]
        [TestSequence(24)]
        public void test_EmptyPassword() {
            SSLContext ctx = setupSSL();
            MQConnection con;

            try {
                con = MQBuilder.BuildConnection(simpleAddress, user, "", ctx);
                con.Close();
            } catch (Exception e) {
                Assert.IsTrue(e == null, "Connect attempt with x.509 ident, name empty password " + e.ToString());
            }

        }

        [Test]
        [TestSequence(25)]
        public void test_BadPassword() {
            SSLContext ctx = setupSSL();
            MQConnection con;

            try {
                con = MQBuilder.BuildConnection(simpleAddress, user, "12345", ctx);
                con.Close();
            } catch (MQException e) {
                Assert.IsTrue(e.getErrorcode() == ErrorCode.EC_NOTAUTHORIZED, "Reconnect with x.509 ident, name wrong password" + e.ToString());
            } catch (Exception e) {
                Assert.IsTrue(false, "Connect attempt with x.509 ident, name wrong password" + e.ToString()); // this is the correct case
            }
        }

        [Test]
        [TestSequence(26)]
        public void test_DefaultSharedContext() {
            SSLContext ctx = setupSSL();
            MQConnection con;

            MQBuilder.ctx = ctx;

            try {
                con = MQBuilder.BuildConnection(simpleAddress, null, null);
                con.Close();
            } catch (Exception e) {
                Assert.IsTrue(e == null, "Connect attempt with x.509 ident null ident" + e.ToString());
            }

            try {
                con = MQBuilder.BuildConnection(simpleAddress, "", "");
                con.Close();
            } catch (Exception e) {
                Assert.IsTrue(e == null, "Connect attempt with x.509 ident empty string" + e.ToString());
            }

            try {
                con = MQBuilder.BuildConnection(simpleAddress, user, null);
                con.Close();
            } catch (Exception e) {
                Assert.IsTrue(e == null, "Connect attempt with x.509 ident name, null password" + e.ToString());
            }

            try {
                con = MQBuilder.BuildConnection(simpleAddress, user, "");
                con.Close();
            } catch (Exception e) {
                Assert.IsTrue(e == null, "Connect attempt with x.509 ident, name empty password" + e.ToString());
            }

            try {
                con = MQBuilder.BuildConnection(simpleAddress, user, "12345");
                con.Close();
            } catch (MQException e) {
                Assert.IsTrue(e.getErrorcode() == ErrorCode.EC_NOTAUTHORIZED, "Reconnect with x.509 ident, name wrong password" + e.ToString());
            } catch (Exception e) {
                Assert.IsTrue(e != null, "Connect attempt with x.509 ident, name wrong password" + e.ToString());
            }
        }
    }
}

File: java/Safmq/SafmqSSLTest.java
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.net.URI;
import java.security.KeyStore;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.Vector;

import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManager;
import javax.net.ssl.TrustManagerFactory;
import javax.net.ssl.X509TrustManager;

import junit.framework.Assert;

import com.safmq.MQBuilder;
import com.safmq.MQConnection;
import com.safmq.MQException;
import com.safmq.Safmq;
import com.safmq.X509Identity;

/**
 * Tests the capability of the Safmq Java client via 
 * 
 * @author Matthew Battey
 */
public class SafmqSSLTest extends SafmqTest {
	String user="npuser";
	String password="54321";
	String description="passwordless login test user";
	String subject = "OU=Testing, O=Safmq Java Test Client";
	String issuer = "C=US, ST=Nebraska, L=Omaha, O=SAFMQ, OU=Sample X509 Certificate, CN=Sample SAFMQ X509 Certificate";

	static X509Certificate certs[] = new X509Certificate[0];

	boolean verified = false;
	
	class MyTrustManager implements X509TrustManager {
		MyTrustManager() { // constructor
			// create/load keystore
		}

		public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
		}

		public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
			verified = true;
		}

		public X509Certificate[] getAcceptedIssuers() {
			return certs;
		}
	}

	public SafmqSSLTest() throws Exception {
		address = new URI("safmqs://admin:@localhost:40001");
		simpleAddress = new URI("safmqs://localhost:40001");
	}

	MQConnection connectToServer(URI uri, String user, String password) {
		try {
			SSLContext ctx = SSLContext.getInstance("SSL");
			TrustManager myTrustManagers[] = {new MyTrustManager()};
			ctx.init(null,myTrustManagers,null);

			MQBuilder.setSSLContext(ctx);
			MQConnection con = MQBuilder.buildConnection(uri, user, password);
			
			Assert.assertTrue("Successfully Connected", con != null);
			return con;
		} catch (Exception e) {
			Assert.assertTrue(e.toString(), e == null);
		}
		return null;
	}
	
	/**
	 * Creates a user and associates an x.509 certificate with that user. Then
	 * attempts to login without a password for that user.
	 */
	public void test_PasswordlessSignon() throws Exception {
		int ec;
		
		
		MQConnection con = connectToServer(address,"","");
		
		ec = con.DeleteUser(user);
		Assert.assertTrue("Delete User", ec == Safmq.EC_NOERROR || ec == Safmq.EC_DOESNOTEXIST);
		
		ec = con.CreateUser(user, password, description);
		Assert.assertTrue("Create User", ec == Safmq.EC_NOERROR);
		
		ec = con.AddUserIdentity(subject, issuer, user);
		Assert.assertTrue("AddUserIdentity", ec == Safmq.EC_NOERROR);
		
		Vector ids = new Vector();
		ec = con.EnumerateUserIdentities(user, ids);
		Assert.assertTrue("EnumerateUserIdentities", ec == Safmq.EC_NOERROR);
		Assert.assertTrue("X.509 ID Count", ids.size() > 0);
		// TODO: Check the results in the ids
		boolean idFound = false;
		for(int x = 0; x < ids.size(); x++) {
			X509Identity id = (X509Identity)ids.get(0);
			if (id.getIssuerDN().equals(issuer) && id.getSubjectDN().equals(subject))
				idFound = true;
		}
		Assert.assertTrue("Identity found in returned X.509 identities", idFound);
		con.Close();
	}
	
	SSLContext setupSSL() throws Exception {
		// Setup the SSL Context using the JKS file created via gen_test_cert in the java/cert directory
		SSLContext ctx = SSLContext.getInstance("SSL");
        KeyManagerFactory   kmf = KeyManagerFactory.getInstance("SunX509", "SunJSSE");
        TrustManagerFactory tmf = TrustManagerFactory.getInstance("SunX509", "SunJSSE");
        KeyStore ks = KeyStore.getInstance("JKS");
        InputStream resrstm = new FileInputStream("cert" + File.separator + "safmq_java_test.jks");
        
        ks.load(resrstm, "".toCharArray());
        kmf.init(ks, "".toCharArray());
        tmf.init(ks);
		ctx.init(kmf.getKeyManagers(),tmf.getTrustManagers(),null);
		
		MQConnection con = connectToServer(address,"","");
		int ec = con.DeleteUser(user);
		Assert.assertTrue("Delete User", ec == Safmq.EC_NOERROR || ec == Safmq.EC_DOESNOTEXIST);
		
		ec = con.CreateUser(user, password, description);
		Assert.assertTrue("Create User", ec == Safmq.EC_NOERROR);
		
		ec = con.AddUserIdentity(subject, issuer, user);
		Assert.assertTrue("AddUserIdentity", ec == Safmq.EC_NOERROR);

		con.Close();
		
		return ctx;
	}
	
	public void test_NullLogin() throws Exception {
		SSLContext ctx = setupSSL();
		MQConnection con;

		try {
			con = MQBuilder.buildConnection(simpleAddress,null,null,ctx);
			con.Close();
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident null ident " + e.toString(), e == null);
		}
	}
	
	public void test_EmptyLogin() throws Exception {
		SSLContext ctx = setupSSL();
		MQConnection con;

		try {
			con = MQBuilder.buildConnection(simpleAddress, "", "", ctx);
			con.Close();
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident empty string " + e.toString(), e == null);
		}
	}

	public void test_NullPassword() throws Exception {
		SSLContext ctx = setupSSL();
		MQConnection con;

		try {
			con = MQBuilder.buildConnection(simpleAddress,user,null,ctx);
			con.Close();
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident name, null password " + e.toString(), e == null);
		}

	}

	public void test_EmptyPassword() throws Exception {
		SSLContext ctx = setupSSL();
		MQConnection con;

		try {
			con = MQBuilder.buildConnection(simpleAddress,user,"",ctx);
			con.Close();
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident, name empty password " + e.toString(), e == null);
		}
		
	}

	public void test_BadPassword() throws Exception {
		SSLContext ctx = setupSSL();
		MQConnection con;


		try {
			con = MQBuilder.buildConnection(simpleAddress,user,"12345",ctx);
			con.Close();
		} catch (MQException e) {
			Assert.assertTrue("Reconnect with x.509 ident, name wrong password" + e.toString(), e.getErrorcode() == Safmq.EC_NOTAUTHORIZED);
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident, name wrong password" + e.toString(), false); // this is the correct case
		}
		
	}

	public void test_DefaultSharedContext() throws Exception {
		SSLContext ctx = setupSSL();
		MQConnection con;

		MQBuilder.setSSLContext(ctx);
		try {
			con = MQBuilder.buildConnection(simpleAddress,null,null);
			con.Close();
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident null ident" + e.toString(), e == null);
		}

		try {
			con = MQBuilder.buildConnection(simpleAddress,"","");
			con.Close();
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident empty string" + e.toString(), e == null);
		}

		try {
			con = MQBuilder.buildConnection(simpleAddress,user,null);
			con.Close();
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident name, null password" + e.toString(), e == null);
		}

		try {
			con = MQBuilder.buildConnection(simpleAddress,user,"");
			con.Close();
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident, name empty password" + e.toString(), e == null);
		}
		
		try {
			con = MQBuilder.buildConnection(simpleAddress,user,"12345");
			con.Close();
		} catch (MQException e) {
			Assert.assertTrue("Reconnect with x.509 ident, name wrong password" + e.toString(), e.getErrorcode() == Safmq.EC_NOTAUTHORIZED);
		} catch (Exception e) {
			Assert.assertTrue("Reconnect with x.509 ident, name wrong password" + e.toString(), e != null);
		}
		

//		con = connectToServer(address,"","");
//		int ec = con.RemoveUerIdentity(subject, issuer);
//		Assert.assertTrue("RemoveUerIdentity", ec == Safmq.EC_NOERROR);
//		con.Close();
	}
}
